### PR TITLE
[SYCL] Add test for kernel program cache

### PR DESCRIPTION
**Pull request by maximdimakov**
This is the test for https://github.com/intel/llvm/pull/5017
Signed-off-by: mdimakov <maxim.dimakov@intel.com>

### CODE REVIEW

File: SYCL/Basic/kernel_program_cache.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: env CreateMultipleRootDevices=2 NEOReadDebugKeys=1 \
+// RUN: SYCL_DEVICE_FILTER="level_zero:gpu" %t.out
+// RUN: env CreateMultipleRootDevices=2 NEOReadDebugKeys=1 \
+// RUN: SYCL_DEVICE_FILTER="opencl:gpu" %t.out
```
Comment (vladimirlaz) [inline SYCL/Basic/kernel_program_cache.cpp]:
```suggestion
// RUN: env CreateMultipleRootDevices=2 NEOReadDebugKeys=1  %GPU_RUN_PLACEHOLDER %t.out
// REQUIRES: (opencl || level_zero) && gpu
```
We have separate runs for level_zero and opencl.

File: SYCL/Basic/kernel_program_cache.cpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  auto kernelID = cl::sycl::get_kernel_id<kernel_name>();
+  auto kernel = bundleObject.get_kernel(kernelID);
+}
+
+int main() {
+  std::vector<device> devices = device::get_devices(info::device_type::gpu);
+  sycl::context context(devices);
+  sycl::queue queue(context, devices[0]);
+  assert(devices.size() == 2);
+  test<empty_functor>(queue, empty_functor{});
```
Comment (vladimirlaz) [inline SYCL/Basic/kernel_program_cache.cpp:40]: Could you add an explicit check that there is no double free? Or the test will crash in this case?

**Review by vladimirlaz — CHANGES_REQUESTED** [review]

File: SYCL/Basic/kernel_program_cache.cpp (hunk 2 above)
Comment (maximdimakov) [reply SYCL/Basic/kernel_program_cache.cpp:40]: The test will crash without changes in https://github.com/intel/llvm/pull/5017

File: SYCL/Basic/kernel_program_cache.cpp (hunk 1 above)
Comment (maximdimakov) [reply SYCL/Basic/kernel_program_cache.cpp]: Changed it

**Review by vladimirlaz — DISMISSED** [review]

File: SYCL/Basic/kernel_program_cache.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: env CreateMultipleRootDevices=2 NEOReadDebugKeys=1  %GPU_RUN_PLACEHOLDER %t.out
```
Comment (vladimirlaz) [inline SYCL/Basic/kernel_program_cache.cpp]:
```suggestion
// RUN: %GPU_RUN_PLACEHOLDER CreateMultipleRootDevices=2 NEOReadDebugKeys=1 %t.out
```

**Pull request closed without merging**